Docs: Reorganize reference v2

NickCrews opened the pull request:
This is a followup per #7141

- Move every datatype-specific API out of the top-level section. eg ibis.array() got moved to collections section
- Now only backend-specific stuff got left in the top-level section. Therefore, I renamed it to "Connections API".

In addition to these high-level changes, also a few tweaks:
- soft deprecate ibis.where
- fix ibis.ifelse docstring (plenty other functions like this are still missing docstrings, but at least the existing ones should be correct)
- import ibis in a doctest to prevent a rendering failure. This was little weird. This wasn't failing before only due to the ordering of this function in the larger API tab. All of the members in a a quartodoc section get squashed into a single .qmd file. Then quarto goes through and evaluates all the docstrings. So if the very first docstring in each API page does `import ibis`, then we are all good. But due to the reordering I did, this no longer happened here, so I had to add this import. At a higher-level, this isn't great, because it means different docstrings share state, and that depends on their ordering in a page. yuck. But maybe not worth it to fix. Possibly we could del all existing variables with some prelude lines, similar to what is done in https://github.com/ibis-project/ibis/pull/7158, but IDK this might all be overkill.

Things that I think should still happen in a followup PR:
- I want it to be more obvious how top level stuff is accessed from ibis. e.g. looking at `now()` in the temporal section, I wish the function signature was `ibis.now()`.
- Make things appear in the correct place where possible, eg `ibis.random()` instead of `ibis.expr.api.random()`.

NickCrews commented:
I just fixed those broken links, but this is ready for review! It is a little hefty, so I appreciate the effort!

cpcloud commented on docs/_quarto.yml:
```diff
@@ -138,9 +130,16 @@ website:
+            - reference/connection.qmd
+            - section: Specific Backend APIs
+              contents:
+                - auto: backends/*.qmd
```
I like our existing backend layout.

Can remove this section and the separate connection section?

How about a how-to that describes how to use `connect`, `set_backend` and `get_backend` instead, that shows their API.

Alternatively, we can have a page that has the base backend API, along with these top level functions.

Either of these can be done in a follow-up.

cpcloud reviewed (changes requested):
If you can remove the changes to the backend layout this LGTM!

NickCrews commented:
I am getting

```
Warning: The computation used to determine source line information from the engine timed out.
Line number information will be unavailable and/or inaccurate.
```
is this to be expected or did I break something?

lostmygithubaccount commented:
> I am getting
> 
> ```
> Warning: The computation used to determine source line information from the engine timed out.
> Line number information will be unavailable and/or inaccurate.
> ```
> 
> is this to be expected or did I break something?

I've never seen this before working w/ Quarto 

I will give a drive-by comment while I'm here that I don't agree w/ hiding the backend docs within the API docs. I understand that it makes sense, as they are API docs to a large extent, but it's useful for users IMO to easily see the list of available backends at a top-level and we can do more platform-specific content in those pages going forward

NickCrews commented:
Ah I think my latest screenshot was confusing. The backend stuff is NOT underneath the rest of the API stuff, it's in its own top level tab as it was before.

cpcloud commented on docs/_quarto.yml:
```diff
@@ -309,6 +267,37 @@ quartodoc:
+            - name: to_sql
+              package: ibis
+              dynamic: true
+            - name: get_backend
```
It looks like this documented in two places: once in `expression-generic` and once in `connection`.

Is that intentional?

NickCrews replied:
Yes. I think this is a functionality that is non-obvious to people, they aren't going to be looking for it. SQL, pandas, Polars, etc have no notion of this. So, I wanted to make it discoverable.

ifelse, on the other hand, people expect to exist and so are going to be looking for it, so that doesn't need to be advertised as much.

I considered the danger of these two places getting out of sync, but I think that is a low chance/impossible, since this is mostly a pointer to the docstring that actually has the content

cpcloud replied:
Generally not in favor of duplicating API documentation like this. I think there are better ways to advertise the API without duplicating its docs, like a blog post or a how-to.

cpcloud commented on docs/_quarto.yml:
```diff
@@ -325,6 +314,23 @@ quartodoc:
+              package: ibis
+              dynamic: true
+            - name: least
+              package: ibis.expr.api
```
`least` and `greatest` should probably be under generic expressions, after all `"a" < "b"` is a well-defined operation for most (all?) backends.

cpcloud replied:
An aside for another PR, maybe we want to aliases these to `max` and `min` for Python users' familiarity. No strong thoughts, just popped into my head.

NickCrews replied:
Oops, yes these should be under generic.

I'm +0 on adding min/max aliases. In that other issue we just removed .where so there is only .ifelse, but this feels warranted. Including the keywords min and max in the docstring would at least make it so I would find it

cpcloud commented on docs/_quarto.yml:
```diff
@@ -325,6 +314,23 @@ quartodoc:
             - DecimalValue
             - name: BooleanValue
               package: ibis.expr.types.logical
+            - name: ifelse
```
I think I would put `ifelse` in generic.

It's closer to the notion of "branching", part of which is of course boolean logic related.

Perhaps a good rule of thumb is:

- If the operation returns or operates on a large variety of types it should go in generic
- Otherwise it should go in the specific docs

I am aware that `ifelse` is defined on bool expressions :)

NickCrews replied:
Sure! I would vote for putting it in both sections, instead of moving it. What do you think?

cpcloud replied:
I think I'd rather have it in single place so that there aren't multiple search results that could be confusing. I see this with sphinx sometimes and I really dislike it.

NickCrews replied:
OK, now it is just in generic in my latest commit stack.

cpcloud reviewed (approved):
This LGTM!

@lostmygithubaccount Can you give a review?

lostmygithubaccount commented:
where is this coming from? should be "Top-level backend API"

<img width="394" alt="image" src="https://github.com/ibis-project/ibis/assets/54814569/d761b681-8f2e-470c-ab70-444f8771054d">

lostmygithubaccount reviewed (approved):
LGTM -- can ignore that nit but if it's easy to fix that'd be nice for consistency

NickCrews commented:
I will push those two fixes!

NickCrews commented:
- fixed the title (I messed up that screenshot, it was a from a stale build of quartodoc, now it is correct)
- moved least, greatest, ifelse to ALSO be in generic. I kept them in the old place too for discoverability.

cpcloud commented on docs/_quarto.yml:
```diff
@@ -111,6 +111,7 @@ website:
       collapse-level: 2
       contents:
         - support_matrix.qmd
+        - reference/connection.qmd
```
Can you move this to be only in the API docs? I feel like we should keep the backend tab to only contain the support matrix and the backends and nothing else. Otherwise it looks cluttered to me.

NickCrews replied:
Sure. Under what section and what should the page be named?

cpcloud replied:
Can you add it this in a follow up PR? In that PR I think you remove this section and let the API docs show up as they usually do under the `Reference` tab.

machow commented:
RE 

> I want it to be more obvious how top level stuff is accessed from ibis. e.g. looking at now() in the temporal section, I wish the function signature was ibis.now().

Just a heads up, in quartodoc you can set `signature_name: full` to show the full path to an object in its signature (e.g. `ibis.now()`). The tricky thing is that you probably don't want the long signatures in methods, so you may need to set it just for those top level functions. (It looks like signature_name is undocumented :/  ).

Here are the docs on setting options: https://machow.github.io/quartodoc/get-started/basic-content.html#setting-default-options

cpcloud commented on docs/_quarto.yml:
```diff
@@ -187,118 +187,76 @@ quartodoc:
-            - name: random
-              dynamic: true
-            - name: range_window
+            - name: set_backend
```
Actually can you add these in a follow up PR? It's hard to see the structural changes here without building the docs.

NickCrews replied:
Not quite sure what you mean by "these"? In my latest push I didn't move anything to underneath the Backends tab. Now just the "top-level" section only contains connection stuff and is renamed "Connection API". Hopefully this is clear enough.

I wish quarto had a preview feature for PRs like this, like readthedocs does... :(

NickCrews replied:
Oh, do you mean just remove commit b3429e9597a0ab73aaddf3a97e4db80345c5f768 from this PR and follow up with that later?

cpcloud replied:
Sorry! By "these" I mean the `set_backend`/`get_backend`/`connect` methods.

NickCrews replied:
I assumed so, so I removed it from the commit stack.

NickCrews replied:
Ah, now I see your clarification, thanks! I think now your concern is rendered useless, as now these methods aren't getting moved. Everything else is just getting moved out, and this section is getting renamed.

NickCrews commented:
OK, I removed https://github.com/ibis-project/ibis/commit/b3429e9597a0ab73aaddf3a97e4db80345c5f768 from this PR because it is a bit unrelated and clutters up the main structure here.

In a followup PR I want to
- include https://github.com/ibis-project/ibis/commit/b3429e9597a0ab73aaddf3a97e4db80345c5f768
- change to `signature_name: full` as @machow describes

NickCrews commented on docs/_quarto.yml:
```diff
@@ -479,6 +476,23 @@ quartodoc:
             - Schema
 
+    - title: Connection APIs
+      package: ibis.expr.operations.udf
```
this top-level package can get removed.

cpcloud replied:
Thanks

cpcloud commented:
I think this is good to go in now!

I moved Connections API to its own section, and fixed a few failing references.

NickCrews commented:
Money, thank you so much for finishing this up and getting it merged!
